### PR TITLE
chore: switch synthetic join key to use ROWKEY

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

This change switching the naming of the synthetic key column created by some joins to use the name `ROWKEY`, or `ROWKEY_1`, `ROWKEY_2` etc if the is a name clash.

The change also contains tidy up and refactoring. Specifically, realizing that the validation `Analyzer` does is shared with pull queries, so it doesn't make sense to move this validation.  Instead, given the synthetic key column is a short to medium term problem only, i.e. it will go away with time, switched the design so that all validation returns to `Analyzer`, except is now won't through on an unknown column is it _might_ be a synthetic key column.  Validation of potential synthetic key columns are handled by the logic model. This works because pull queries don't (yet) have this problem.

 - There was duplicate logic between `AstSantizer` and `Analayzer` around reporting ambiguous and unknown columns.
    - Changed so that, for now, `AstSanitzer` reports ambiguous columns and `Analyzer` reports unknown columns. Not perfect, but better than both attempting to do both.
 - It puts back the validation of Selects in `Analyser` and enhanced it and `ColumnReferenceValidator` to know when a synthetic key column is possible and to not throw on one. False positives are possible, but are caught by the logical model.
 - Simplified the logical model to only worry about validating synthetic key columns in projections.
 - Split `RepartitionNode` into classes specific for its two use-cases.

#### Reviewing notes:

Commits split in the usual way.

### Testing done 

usual

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/ColumnReferenceValidator.java [hunk 1]
```diff
@@ -85,46 +83,55 @@ public Void visitQualifiedColumnReference(
...
 
-      if (sourcesWithField.size() > 1) {
-        final String possibilities = sourcesWithField.stream()
-            .map(source -> new QualifiedColumnReferenceExp(source, name))
-            .map(Objects::toString)
-            .sorted()
-            .collect(Collectors.joining(", "));
+      final SourceName source = colRef.maybeQualifier()
+          .orElseGet(() -> {
+            // AstSanitizer should catches ambiguous columns
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/ColumnReferenceValidator.java:105]:
```suggestion
            // AstSanitizer should catch ambiguous columns
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/ColumnReferenceValidator.java [hunk 2]
```diff
@@ -85,46 +83,55 @@ public Void visitQualifiedColumnReference(
...
-      final String loc = location
-          .map(Objects::toString)
-          .map(text -> text + ": ")
-          .orElse("");
+      if (colRef instanceof QualifiedColumnReferenceExp) {
+        // Synthetic join columns can't be qualified, as they don't belong to any source
+        return false;
+      }
+
+      if (!clauseType.equals("SELECT")) {
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/ColumnReferenceValidator.java:128]: seems like if we're checking equality it should be made either an enum or a constant

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PreJoinRepartitionNode.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.planner.plan;
+
+import io.confluent.ksql.execution.expression.tree.Expression;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+
+public class PreJoinRepartitionNode extends RepartitionNode {
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PreJoinRepartitionNode.java:21]: would be nice to javadoc

File: ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/ColumnReferenceValidator.java (hunk 2 above)
Comment (big-andy-coates) [reply ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/ColumnReferenceValidator.java:128]: Normally I'd agree, but it's only short term code and there are tests to ensure it does what we need.

**Pull request merged**